docs: drop meta info for deprecated functions

justinmk opened the pull request:
A central tenet of our deprecation strategy https://github.com/neovim/neovim/issues/31376 is that deprecated functions must have zero footprint in docs, auto-completion, and meta structures. The upside is that it greatly reduces the urge to remove the actual *implementation*, so legacy programmatic consumers can worry less about migrating.

(Note: formal "deprecation strategy" proposal is still WIP, but the above is not debatable.)

todo:

- formalize api-contract for dict params; also allow "widening" to Object. [ref](https://github.com/neovim/neovim/pull/30844#issuecomment-2424982640)

lewis6991 commented:
> and meta structures.

I'm against this. The meta structures are useful to communicate to a user that a function is deprecated (with `@deprecated`) with an optional message of when. Without this it can be confusing if the function exists or is spelled incorrectly.

justinmk commented:
> The meta structures are useful to communicate to a user that a function is deprecated (with `@deprecated`) with an optional message of when.

I will consider adding the signatures and `@deprecated` note, but still removing the rest of the docs.

> > The upside is that it greatly reduces the urge to remove the actual implementation
> 
> You remove the urge by incorporating removal into the deprecation strategy.

That is the inverse approach. Note that both approaches are part of the strategy, but reducing the urgency of removals will buy time and reduce work, while also allowing more experimentation.

lewis6991 commented:
> I will consider adding the signatures and @deprecated note, but still removing the rest of the docs.

Fine with that. That'll at least maintain any deprecated diagnostics.

justinmk commented:
> how it allows more experimentation?

Reducing the cost of potential deprecations reduces friction for introducing (some classes of) new interfaces in the first place.

lewis6991 commented:
I'm not sure how this reduces the cost of deprecations. The usual cost of deprecations is the ongoing maintenance of said deprecated functions.

lewis6991 commented:
This change introduces a high cost to users who will lack the documentation required to migrate code.

The existence of the documentation and added surface (3) is a low cost provided the functions are marked accordingly. It's not like users read the full documentation from start to finish and force themselves to remember it. It's simply reference material.

justinmk committed:
docs: no meta info for deprecated functions

A central tenet of our deprecation strategy is that deprecated functions
must have zero footprint in docs, auto-completion, and meta structures.

justinmk commented:
> This change introduces a high cost to users who will lack the documentation required to migrate code.

It introduces some cost there, but the docs are still available in various places, just not in our main tree. And removing something entirely is a higher cost, which can be avoided for cases that take this approach.

lewis6991 commented:
> which can be delayed or avoided for cases that take this approach.

Why? I don't understand how removing the docs has any bearing whatsoever. The meta information isn't even in the documentation, it's just something that exists to be consumed by tools so they have stronger semantic understanding. The cost to keep that is zero, especially in this case where it's generated.

If you don't want to remove these functions then just say that. Otherwise what's stopping someone removing them in a follow-up?

mfussenegger commented:
I second @lewis6991 stance here.

From API consumer perspective, a common workflow is:

- You see a function marked as deprecated (requires the metainfo for luals to do that)
- You read the docs (hoping it contains a pointer to what to do about the deprecation)
- You change the call accordingly.

If we remove the metainfo it reduces the chance that plugin authors discover deprecations early. If we remove the docs, it makes it harder for them to figure out how to act.

This is _increasing_ friction for deprecations, not lowering it from a consumer point of view.

And from a maintainer point of view leaving both in has almost no costs. Docs are not coupled to anything else - there is no interaction with other systems that would cause pain (unless the edge case of changing the doc system)

The cognitive load of seeing a deprecated function is also not that high - as long as it's clear that the function is documented.

justinmk commented:
We can make this conditional on the api-level where a function was deprecated.

For example in this very PR, notice how 20-year-old vimscript functions are in the meta info. Those things exist only for "ABI" compat. Removing most human-visible evidence allows us to keep things around longer for programmatic compat, while still achieving the other benefits of deprecation.

By making this conditional on the *age* of the deprecation, this allows us to buy time.

There is no good reason that `buffer_exists` should appear in completion or documentation.

lewis6991 commented:
There is no "buying time" here. You either keep the docs or not. Neither of those decisions has any bearing of how long the depreciations continue to exist for. They exist for as long as we allow them to.

In fact I think they may be out of time anyway... 

Have you considered just removing these? If they are not valuable enough to keep documentation for, then they are not valuable enough to keep. Their implementation may be tiny, but their usage is likely zero (or very close to).

lewis6991 commented on scripts/gen_eval_files.lua:
```diff
@@ -345,8 +345,10 @@ local function render_api_meta(_f, fun, write)
-    if pdesc then
+    local pdesc = not fun.deprecated and p[3] or nil
+    if fun.deprecated then -- luacheck: ignore 542
+      -- Skip @param description for deprecated functions.
```
What does this special casing achieve?
